### ISSUE TITLE
Some emojis are not rendered in color.

### ISSUE THREAD

**Issue by k4l1brx** [alacritty/alacritty#4692]
### Problem:

Some emojis are not rendered in color.

![](https://i.imgur.com/VzXLwaA.png)
The left pane is alacritty, the right pane is kitty. 

The emoji file:
```
 😀 😃 😄 😁 😆 😅 😂 🤣 🥲 ☺️ 😊 😇 🙂 🙃 😉 😌 😍 🥰 😘 😗 😙 😚 😋 😛 😝 😜 🤪 🤨 🧐 🤓 😎 🥸 🤩 🥳 😏 😒 😞 😔 😟 😕 🙁 ☹️ 😣 😖 😫 😩 🥺 😢 😭 😤 😠 😡 🤬 🤯 😳 🥵 🥶 😱 😨 😰 😥 😓 🤗 🤔 🤭 🤫 🤥 😶 😐 😑 😬 🙄 😯 😦 😧 😮 😲 🥱 😴 🤤 😪 😵 🤐 🥴 🤢 🤮 🤧 😷 🤒 🤕 🤑 🤠 😈 👿 👹 👺 🤡 💩 👻 💀 ☠️ 👽 👾 🤖 🎃 😺 😸 😹 😻 😼 😽 🙀 😿 😾 
🧳 🌂 ☂️ 🧵 🪡 🪢 🧶 👓 🕶 🥽 🥼 🦺 👔 👕 👖 🧣 🧤 🧥 🧦 👗 👘 🥻 🩴 🩱 🩲 🩳 👙 👚 👛 👜 👝 🎒 👞 👟 🥾 🥿 👠 👡 🩰 👢 👑 👒 🎩 🎓 🧢 ⛑ 🪖 💄 💍 💼 
```

### System

OS: OpenSUSE Tumbleweed
Version: alacritty 0.7.1
Linux: X11, Xmonad, compositor: picom v8.2

### My alacritty config:
I ran without it and the problem was still there. 
```
env:
  TERM: alacritty

window:
  dynamic_title: true
  # Window dimensions (changes require restart)
  #
  # Specified in number of columns/lines, not pixels.
  # If both are `0`, this setting is ignored.
  #dimensions:
  #  columns: 0
  #  lines: 0

  # Window position (changes require restart)
  #
  # Specified in number of pixels.
  # If the position is not set, the window manager will handle the placement.
  #position:
  #  x: 0
  #  y: 0

  # Window padding (changes require restart)
  #
  # Blank space added around the window in pixels. This padding is scaled
  # by DPI and the specified value is always added at both opposing sides.
  padding:
    x: 2
    y: 2

  # Spread additional padding evenly around the terminal content.
  dynamic_padding: false

  # Window title
  title: Alacritty

  # Window class (Linux/BSD only):
  class:
    # Application instance name
    instance: Alacritty
    # General application class
    general: Alacritty

  # GTK theme variant (Linux/BSD only)
  #
  # Override the variant of the GTK theme. Commonly supported values are `dark` and `light`.
  # Set this to `None` to use the default theme variant.
  gtk_theme_variant: None

scrolling:
  # Maximum number of lines in the scrollback buffer.
  # Specifying '0' will disable scrolling.
  history: 10000

  # Number of lines the viewport will move for every line scrolled when
  # scrollback is enabled (history > 0).
  multiplier: 1
  
bell:
  animation: EaseOutExpo
  duration: 0
  color: '#ffffff'
background_opacity: 1.0

selection:
  semantic_escape_chars: ",│`|:\"' ()[]{}<>\t"

live_config_reload: true
working_directory: None
mouse:
  double_click: { threshold: 300 }
  triple_click: { threshold: 300 }
  url:
    launcher:
      program: xdg-open
      args: []
    modifiers: None

mouse_bindings:
  - { mouse: Middle, action: Copy }

key_bindings:
  - { key: P,        mods: Control,       action: Paste            }
  - { key: Insert,   mods: Shift,         action: Paste            }
  - { key: Slash,    mods: Control,       chars: "gc"              }
  - { key: Y,        mods: Control,       action: Copy             }
  - { key: Key0,     mods: Control,       action: ResetFontSize    }
  - { key: Equals,   mods: Control,       action: IncreaseFontSize }
  - { key: Plus,     mods: Control,       action: IncreaseFontSize }
  - { key: Minus,    mods: Control,       action: DecreaseFontSize }
  - { key: Return,   mods: Control|Shift, action: SpawnNewInstance }

debug:
  render_timer: false

  persistent_logging: false

  log_level: Warn
  print_events: false
font:
    size: 10
    normal:
        family: FiraMono Nerd Font 
        style: Medium
    bold:
        family: FiraMono Nerd Font 
        style: Bold
    italic:
        family: FiraMono Nerd Font 
        style: Italic
  # Offset is the extra space around each character. `offset.y` can be thought of
  # as modifying the line spacing, and `offset.x` as modifying the letter spacing.
    offset:
        # x: -1
        y: -1

    glyph_offset:
        x: 0
        y: 0

draw_bold_text_with_bright_colors: false
special:
    background: "#1d2021"
    foreground: "#d5c4a1"
    cursor: "#d5c4a1"

colors:
    color0: "#1d2021"
    color1: "#fb4934"
    color2: "#b8bb26"
    color3: "#fabd2f"
    color4: "#83a598"
    color5: "#d3869b"
    color6: "#8ec07c"
    color7: "#d5c4a1"
    color8: "#665c54"
    color9: "#fb4934"
    color10: "#b8bb26"
    color11: "#fabd2f"
    color12: "#83a598"
    color13: "#d3869b"
    color14: "#8ec07c"
    color15: "#fbf1c7"
```

### Logs

`alacritty -vv`

```
Created log file at "/tmp/Alacritty-3891.log"
[2021-01-12 15:51:52.631407230] [INFO ] [alacritty] Welcome to Alacritty
[2021-01-12 15:51:52.631667324] [INFO ] [alacritty] Configuration files loaded from:
                                                      "/home/hts/.config/alacritty/alacritty.yml"
[2021-01-12 15:51:52.763111148] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Medium, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:51:52.784040101] [DEBUG] [alacritty] Estimated DPR: 1.0833333333333333
[2021-01-12 15:51:52.784116130] [DEBUG] [alacritty] Estimated window size: None
[2021-01-12 15:51:52.784153676] [DEBUG] [alacritty] Estimated cell size: 8 x 19
[2021-01-12 15:51:52.808373198] [INFO ] [alacritty] Device pixel ratio: 1.0833333333333333
[2021-01-12 15:51:52.817186017] [INFO ] [alacritty] Initializing glyph cache...
[2021-01-12 15:51:52.902613184] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Medium, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:51:53.005972761] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:51:53.112460108] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:51:53.214890392] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:51:53.241545138] [INFO ] [alacritty] ... finished initializing glyph cache in 0.424309933s
[2021-01-12 15:51:53.241654300] [INFO ] [alacritty] Cell size: 8 x 19
[2021-01-12 15:51:53.241675842] [INFO ] [alacritty] Padding: 2 x 2
[2021-01-12 15:51:53.241689080] [INFO ] [alacritty] Width: 800, Height: 600
[2021-01-12 15:51:53.244523163] [INFO ] [alacritty] PTY dimensions: Line(31) x Column(99)
[2021-01-12 15:51:53.258975840] [INFO ] [alacritty] Initialisation complete
[2021-01-12 15:51:53.259624844] [DEBUG] [alacritty_terminal] New num_cols is 170 and num_lines is 38
[2021-01-12 15:51:53.263300211] [INFO ] [alacritty] Padding: 2 x 2
[2021-01-12 15:51:53.263400047] [INFO ] [alacritty] Width: 1366, Height: 733
[2021-01-12 15:51:53.373953574] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Bold, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:51:58.568755057] [DEBUG] [alacritty_terminal] New num_cols is 84 and num_lines is 38
[2021-01-12 15:51:58.569341150] [INFO ] [alacritty] Padding: 2 x 2
[2021-01-12 15:51:58.569453576] [INFO ] [alacritty] Width: 679, Height: 729
[2021-01-12 15:52:04.536026433] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Book, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:52:04.537989356] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:52:04.539132313] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_HINTING | FORCE_AUTOHINT | COLOR, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:52:04.556818246] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Medium, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:52:04.559734945] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:52:04.568050669] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Medium, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:52:04.568523194] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_HINTING | NO_BITMAP | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
[2021-01-12 15:52:04.571166158] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: NO_HINTING | FORCE_AUTOHINT, render_mode: "Normal", lcd_filter: 0 }
```

**Comment by chrisduerr** [comment]
Alacritty uses font fallback to determine which font it should pick. If you have a different font with emojis at a higher priority than your colored emoji font, this is what you'll get. You should either uninstall that other font or change its priority.

**Comment by misaflo** [comment]
Hello,

I solved this problem with adding this file: `~/.config/fontconfig/fonts.conf`

    <?xml version="1.0"?>
    <!DOCTYPE fontconfig SYSTEM "fonts.dtd">
    <fontconfig>
    <alias>
       <family>Monospace</family>
       <prefer>
         <family>Monospace</family>
         <family>Noto Color Emoji</family>
       </prefer>
     </alias>
    </fontconfig>

**Comment by XReaper95** [comment]
For anyone having this problem in Arch ( tested in `6.1.41-1-MANJARO XFCE`), that last config didn't work for me. I used this one instead:
```
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
  <alias>
    <family>serif</family>
    <prefer>
      <family>Joy Pixels</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>
  <alias>
    <family>sans-serif</family>
    <prefer>
      <family>Joy Pixels</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>
  <alias>
    <family>monospace</family>
    <prefer>
      <family>Noto Sans Mono</family>
      <family>Joy Pixels</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>
  <alias>
    <family>sans</family>
    <prefer>
      <family>Joy Pixels</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>
</fontconfig>
```

**Comment by XavierFigari** [comment]
> Alacritty uses font fallback to determine which font it should pick. If you have a different font with emojis at a higher priority than your colored emoji font, this is what you'll get. You should either uninstall that other font or change its priority.

How do you define fallback fonts in the .toml configuration file ? This is not documented [here](https://alacritty.org/config-alacritty.html).

**Comment by chrisduerr** [comment]
> How do you define fallback fonts in the .toml configuration file ?

You don't, it's defined by operating system configuration through fontconfig.

**Comment by XavierFigari** [comment]
Got it, thanks !

There are actually many threads related to this.

I ended up creating a fontconfig configuration file in my ~/.config/fontconfig directory, as described [here](https://github.com/alacritty/alacritty/issues/957#issuecomment-1553255963) :

File : `~/.config/fontconfig/fonts.conf`
```xml
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "urn:fontconfig:fonts.dtd">
<fontconfig>
<alias>
   <family>Noto Sans Mono</family>
   <prefer>
     <family>Noto Sans Mono</family>
     <family>Noto Color Emoji</family>
   </prefer>
 </alias>
</fontconfig>
```

I think I then ran `fc-cache -fv`, restarted Alacritty, and voila!

The approach is understandable, but a bit tedious if you do not manipulate fontconfig every day. For the first time user of Alacritty, it takes a long time to find the way to display colored emojis correctly, while some terminals like Guake do it out of the box (I don't know how).

I see that some suggestions have already been made to implement font fallback in Alacritty, I would support that. But I don't know how much effort it would take, and I don't know Rust, so I can easily understand why it's not a top priority.

Thanks again for this great terminal, which might become my preferred terminal now 🙂 .
